annotated.go

```diff
@@ -89,3 +90,268 @@ func (a Annotated) String() string {
 }
+
+type annotations struct {
+	ParamTags  []string
+	ResultTags []string
+}
+
+// Annotation can be passed to Annotate(f interface{}, anns ...Annotation)
+// for annotating the parameter and result types of a function.
+type Annotation interface {
+	apply(*annotations) error
+	getAnnotatedType(reflect.Type) []reflect.Type
+}
+
+type annotationError struct {
+	target interface{}
+	err    error
+}
+
+func (e *annotationError) Error() string {
+	return e.err.Error()
+}
+
+type paramTags struct {
+	tags []string
+}
+
+func (pt paramTags) apply(ann *annotations) error {
+	if len(ann.ParamTags) > 0 {
+		return fmt.Errorf("cannot apply more than one line of ParamTags")
+	}
+	ann.ParamTags = pt.tags
+	return nil
+}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.In
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (pt paramTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedParams := []reflect.StructField{{
+		Name:      "In",
+		Type:      reflect.TypeOf(In{}),
+		Anonymous: true,
+	}}
+
+	for i := 0; i < fType.NumIn(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.In(i),
+		}
+		if i < len(pt.tags) {
+			structField.Tag = reflect.StructTag(pt.tags[i])
+		}
+		annotatedParams = append(annotatedParams, structField)
+	}
+
+	paramTypes := []reflect.Type{reflect.StructOf(annotatedParams)}
+	return paramTypes
+}
+
+// ParamTags is an Annotation that annotates the parameter(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional parameter.
+func ParamTags(tags ...string) Annotation {
+	p := paramTags{tags}
+	return p
+}
+
+type resultTags struct {
+	tags []string
+}
+
+func (rt resultTags) apply(ann *annotations) error {
+	if len(ann.ResultTags) > 0 {
+		return fmt.Errorf("cannot apply more than one line of ResultTags")
+	}
+	ann.ResultTags = rt.tags
+	return nil
+}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.Out
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (rt resultTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	errInterface := reflect.TypeOf((*error)(nil)).Elem()
+	numAnnotated := 0
+
+	for i := 0; i < fType.NumOut(); i++ {
+		// guard against error results
+		if fType.Out(i).Implements(errInterface) {
+			continue
+		}
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", numAnnotated),
+			Type: fType.Out(i),
+		}
+		if numAnnotated < len(rt.tags) {
+			structField.Tag = reflect.StructTag(rt.tags[numAnnotated])
+		}
+		numAnnotated++
```

Comment:
Why is numAnnotated different from `i` here? If this is to handle the `error` case, I guess we should decide what we want the user code to look like. I assumed,

```
fx.Annotate(func(...) (Thing, error), fx.ResultTags(`name:"foo"`))
```

If there's a result after error, then,

```
fx.Annotate(func(...) (Thing, error, AnotherThing), fx.ResultTags(`name:"foo"`, "", `name:"bar"`))
```

In which case, `i` should be used, not `numAnnotated`. I think.

Of course, this might all be moot because I _think_ (but have not verified) that Fx will reject constructors that return errors as anything but their last result.

Reply:
Oh, I suppose I could make the resulttag be an empty string for an error. Yeah, this is the handle the exact case you have there, but I was thinking the user would then need to do:
```
fx.Annotate(func(...) (Thing, error, AnotherThing), fx.ResultTags(`name:"foo"`, `name:"bar"`))
```
instead, which is why I have `numAnnotated` instead of `i`. 

But I suppose the user could just pass in an empty string here. Hm.. I'm not sure what would be more intuitive thing for the user to do.

Reply:
it's been updated to expect an empty string for the error result :) Thanks for the feedback!